rust/src/pool/mod.rs

```diff
@@ -0,0 +1,403 @@
+use std::{
+    ops::DerefMut,
+    sync::{Arc, LockResult, RwLock},
+};
+
+use crate::error::Error;
+
+/// An abstraction for a pool of thread-safe elements
+pub trait Pool {
+    /// The key type used to identify entries in the pool.
+    type Key;
+    /// The type of elements indexed by the pool.
+    type Type;
+
+    /// Retrieves an entry from the pool.
```

Comment:
```suggestion
    /// Retrieves an entry from the pool.
    /// The concrete type returned by the get operation may not be the concrete type `Self::Type` but instead 
    /// a wrapper type which behaves like `Self::Type`. This abstraction allows for the pool to store complex types which track 
    /// the lifetime of stored objects, enabling nested catching. 
```

This needs some work, as the returned is the lock object, which wraps the internal type... you get the idea